### ISSUE TITLE
New jobs don't start on the top

### ISSUE THREAD

**Issue by gallardoalba** [galaxyproject/galaxy#12605]
Describe the bug
When launching a new job, it appears at the bottom of the history.

Galaxy Version and/or server at which you observed the bug
Galaxy Version: 21.09 (usegalaxy.org). 

To Reproduce
Use the Remove_beginning1 tool on this history: https://usegalaxy.org/u/gallardoalba/h/imported-rna-seq-analysis

Expected behavior
The job starts on the top.

Actual behavior
The job starts at the bottom.

Screenshots
https://user-images.githubusercontent.com/26678117/135470633-03faed3d-05b6-4e7c-be11-920bbcbca530.mp4

Additional context
Discovered during 21.09 release testing.

**Comment by mvdbeek** [comment]
That one is extremely weird 😨. I can reproduce this using the imported history, except the hid is 4 for me.

**Comment by gallardoalba** [comment]
> That one is extremely weird fearful. I can reproduce this using the imported history, except the hid is 4 for me.

I guess that if you repeat it again, it will appear in the 5th position.

**Comment by ic4f** [comment]
Verified too. On this history only - not on other histories. I have no idea what might be causing it, but I'll try to at least narrow it down for now..

**Comment by ic4f** [comment]
It always starts as hid=4 and then increments it by 1. So, 4 is the first non-duplicate hid in the history. I'm sorry - I'm so ignorant of Galaxy's day-to-day usage, I don't even know if having duplicate hids is normal. And if it is, how can I reproduce that? (and if no, then that could be a pointer to where the problem is)

**Comment by ic4f** [comment]
So, yes. The hid of that history in the database is 4, which is clearly incorrect. I'll poke around to see if I can find what conditions may have caused this.

**Comment by ic4f** [comment]
@gallardoalba Was this history originally imported from somewhere? I'm hoping to trace this to where the hid was set to 1: there are 3 duplicate hids, all correspond to the "remove beginning" tool, which only means that when that tool was first executed, the hid on the history was 1. Given that the history already had 20 datasets, the hid should have been 21, so, I suspect, it is *not* a result of a race condition (i may be wrong though). More likely, given that this history was originally imported, hid was not set at import time, so it (re)started at 1 (just a wild guess at this point).

**Comment by mvdbeek** [comment]
I think that's exactly what is happening. 
https://github.com/galaxyproject/galaxy/blob/bf3b6cd6f0e4fe0ddcdc0012c568d6291a440608/lib/galaxy/model/__init__.py#L2543 should be where we update the hid counter.

Probably worth checking what the original hid counter of the history is at right now. I also think `new_history.hid_counter = self.hid_counter` is pretty prone to race conditions in general, it might be a good idea to set `new_history.hid_counter` to the max hid of items that are going to be copied (which leaves the target history open to races, but since this happens in a request cycle I think that is ok for now).

**Comment by gallardoalba** [comment]
> @gallardoalba Was this history originally imported from somewhere? I'm hoping to trace this to where the hid was set to 1: there are 3 duplicate hids, all correspond to the "remove beginning" tool, which only means that when that tool was first executed, the hid on the history was 1. Given that the history already had 20 datasets, the hid should have been 21, so, I suspect, it is _not_ a result of a race condition (i may be wrong though). More likely, given that this history was originally imported, hid was not set at import time, so it (re)started at 1 (just a wild guess at this point).

Yes, it was the original history. It seems to have been truncated during the import process (I think that it is related to the migration of the new storage) @mvdbeek @ic4f.

**Comment by ic4f** [comment]
The current `hid_counter` of this history in the database is 4. 

I've looked in the db again, and, I think, it appears that the data has been indeed truncated at import, as per @gallardoalba . The original history (id=5251678) has 40 datasets, 38 out of those are the initial items (version=1, created at the same moment `2021-09-25 23:05:36.*`). The first 20 correspond to the first 20 datasets in the imported history we are looking at (id=5255858). The `hid`s in both are 1 through 20. At that point the imported history diverges: there are 3 new datasets with `hid`s 1 through 3, which is why the `hid_counter` in the history table is 4.

```
galaxy_main=> select id, dataset_id, create_time, update_time, hid, name, version from history_dataset_association where history_id = 5255858 order by create_time;
    id    | dataset_id |        create_time         |        update_time         | hid |                         name                         | version 
----------+------------+----------------------------+----------------------------+-----+------------------------------------------------------+---------
 77169063 |   61558601 | 2021-09-28 11:57:22.519702 | 2021-09-30 19:20:57.726281 |   1 | GSM461180_1.fastqsanger                              |       5
 77169064 |   61558677 | 2021-09-28 11:57:22.519713 | 2021-09-30 19:20:57.726285 |   2 | GSM461177_1.fastqsanger                              |       5
...[rows 3-19]...
 77169082 |   61560194 | 2021-09-28 11:57:22.519805 | 2021-09-30 19:20:57.726315 |  20 | Trimmomatic on GSM461177_1.fastqsanger (R2 unpaired) |       8
 77427225 |   62183124 | 2021-09-30 13:58:29.016927 | 2021-09-30 19:20:57.726316 |   1 | Remove beginning on data 20                          |       6
 77427235 |   62183134 | 2021-09-30 13:59:15.138351 | 2021-09-30 19:20:57.726318 |   2 | Remove beginning on data 20                          |       6
 77427375 |   62183274 | 2021-09-30 14:05:13.583138 | 2021-09-30 19:20:57.726319 |   3 | Remove beginning on data 20                          |       6
(23 rows)
```

I suggest we consider this particular instance to be a glitch in the matrix, but keep the issue open, removing the 21.09 milestone. I'll look at how we deal with hid assignment to see if we can tighten it some (+ will try @mvdbeek 's suggestion)

**Comment by gallardoalba** [comment]
This is the original history @ic4f https://usegalaxy.org/u/rajeevnepal/h/unnamed-history. I forgot to share it before.
